catalog/app/containers/Bucket/PackageUpdateDialog.js

```diff
@@ -25,7 +26,460 @@ import * as validators from 'utils/validators'
+      FilesEntry.case({
+        File: () => false,
+        Dir: R.propEq('name', current),
+      }),
+    )
+    if (existingDir) {
+      restEntries = R.without([existingDir], entries)
+      baseDir = existingDir
+    }
+    inserted = insertIntoDir(rest, file, baseDir)
+  }
+  const getOrder = FilesEntry.case({
+    Dir: (d) => [0, d.name],
+    File: (f) => [1, f.name],
+  })
+  const tree = R.sortBy(getOrder, [inserted, ...restEntries])
+  return tree
+}
+
+const insertIntoDir = (path, file, dir) => {
+  const { name, children } = FilesEntry.Dir.unbox(dir)
+  const newChildren = insertIntoTree(path, file, children)
+  const type = newChildren
+    .map(FilesEntry.case({ Dir: R.prop('type'), File: R.prop('type') }))
+    .reduce((acc, entryType) => (acc === entryType ? acc : 'modified'))
+  return FilesEntry.Dir({ name, type, children: newChildren })
+}
+
+const dissocBy = (fn) =>
+  R.pipe(
+    R.toPairs,
+    R.filter(([k]) => !fn(k)),
+    R.fromPairs,
+  )
+
+/*
+// TODO: use tree as the main data model / source of truth?
+state type: {
+  existing: { [path]: file },
+  added: { [path]: file },
+  // TODO: use Array or Set?
+  deleted: { [path]: true },
+}
+*/
+
+const handleFilesAction = FilesAction.case({
+  Add: ({ files, prefix }) => (state) =>
+    files.reduce((acc, file) => {
+      const path = (prefix || '') + PD.getNormalizedPath(file)
+      return R.evolve(
+        {
+          added: R.assoc(path, file),
+          deleted: R.dissoc(path),
+        },
+        acc,
+      )
+    }, state),
+  Delete: (path) => R.evolve({ deleted: R.assoc(path, true) }),
+  // add all descendants from existing to deleted
+  DeleteDir: (prefix) => ({ existing, added, deleted }) => ({
+    existing,
+    added,
+    deleted: R.mergeLeft(
+      Object.keys(existing).reduce(
+        (acc, k) => (k.startsWith(prefix) ? { ...acc, [k]: true } : acc),
+        {},
+      ),
+      deleted,
+    ),
+  }),
+  Revert: (path) => R.evolve({ added: R.dissoc(path), deleted: R.dissoc(path) }),
+  // remove all descendants from added and deleted
+  RevertDir: (prefix) =>
+    R.evolve({
+      added: dissocBy(R.startsWith(prefix)),
+      deleted: dissocBy(R.startsWith(prefix)),
+    }),
+  Reset: (_, { initial }) => () => initial,
+})
+
+const computeEntries = ({ added, deleted, existing }) => {
+  const existingEntries = Object.entries(existing).map(([path, { size }]) => {
+    if (path in deleted) {
+      return { type: 'deleted', path, size }
+    }
+    if (path in added) {
+      const a = added[path]
+      return { type: 'modified', path, size: a.size }
+    }
+    return { type: 'unchanged', path, size }
+  })
+  const addedEntries = Object.entries(added).reduce((acc, [path, { size }]) => {
+    if (path in existing) return acc
+    return acc.concat({ type: 'added', path, size })
+  }, [])
+  const entries = [...existingEntries, ...addedEntries]
+  return entries.reduce((children, { path, ...rest }) => {
+    const parts = path.split('/')
```

Comment:
I don't know, Windows has `/` here too? If not, there is `path.sep`, but I don't know if it works in browser correctly.

Reply:
i've tried and it works as expected on windows, probably bc all the directory reading and path normalization stuff is handled by `react-dropzone` / `file-selector` packages